diagnostics/include/diagnostics_server.h

```diff
@@ -0,0 +1,166 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+// This file contains a tcp server listening on localhost that handles the messages in protocol.h.
+
+#pragma once
+
+#include <atomic>
+#include <cassert>
+#include <chrono>
+#include <cstring>
+#include <future>
+#include <sstream>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <thread>
+#include <unistd.h>
+
+#include "errnoString.hpp"
+#include "protocol.h"
+
+using concordUtils::errnoString;
+
+namespace concord::diagnostics {
+
+static constexpr uint16_t PORT = 6888;
+static constexpr int BACKLOG = 5;
+static constexpr size_t MAX_INPUT_SIZE = 1024;
+
+// Returns a successfully read line as a string.
+// Throws a std::runtime_error on error.
+std::string readline(int sock) {
+  std::array<char, MAX_INPUT_SIZE> buf;
+  buf.fill(0);
+  int count = 0;
+  auto start = std::chrono::steady_clock::now();
+  auto timeout = std::chrono::microseconds(999999);
+  auto remaining = timeout;
+  while (true) {
+    fd_set read_fds;
+    FD_ZERO(&read_fds);
+    FD_SET(sock, &read_fds);
+    timeval tv;
+    tv.tv_sec = 0;
+    tv.tv_usec = remaining.count();
+    auto rv = select(sock + 1, &read_fds, NULL, NULL, &tv);
+    if (rv == 0) {
+      throw std::runtime_error("timeout");
+    }
+    if (rv < 0 && errno == EINTR) continue;
```

Comment:
What makes `EINTR` special in this case?

Reply:
EINTR is not actually an error. It can happen due to a signal interrupting any blocking syscall. I'm not sure how often this happens anymore, but I've seen and written this code many times where EINTR is checked and ignored. 

See https://github.com/tokio-rs/mio/issues/494 and https://github.com/tokio-rs/mio/issues/16